### PR TITLE
Add HealthIndicator for Hazelcast

### PR DESCRIPTION
**Pull request by nosan**
see gh-17498

### CODE REVIEW

File: spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/hazelcast/HazelcastHealthIndicator.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	protected void doHealthCheck(Health.Builder builder) {
+		HazelcastInstance hazelcast = this.hazelcast;
+		hazelcast.executeTransaction((context) -> builder.up().withDetails(getDetails(hazelcast)));
+	}
+
+	private static Map<String, Object> getDetails(HazelcastInstance hazelcast) {
+		Map<String, Object> details = new LinkedHashMap<>();
+		details.put("name", hazelcast.getName());
+		Endpoint endpoint = hazelcast.getLocalEndpoint();
+		if (endpoint instanceof Member) {
```
Comment (snicoll) [inline spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/hazelcast/HazelcastHealthIndicator.java]: I would stick to `name` and `uuid` for now. Not denying that extra information might be useful but I'd start with a lighter surface area for a start.

File: spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/hazelcast/HazelcastHealthIndicator.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	public HazelcastHealthIndicator(HazelcastInstance hazelcast) {
+		super("Hazelcast health check failed");
+		Assert.notNull(hazelcast, "HazelcastInstance must not be null");
+		this.hazelcast = hazelcast;
+	}
+
+	@Override
+	protected void doHealthCheck(Health.Builder builder) {
+		HazelcastInstance hazelcast = this.hazelcast;
+		hazelcast.executeTransaction((context) -> builder.up().withDetails(getDetails(hazelcast)));
```
Comment (snicoll) [inline spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/hazelcast/HazelcastHealthIndicator.java]: Curious as why this needs to run in a transaction.

File: spring-boot-project/spring-boot-actuator/src/test/java/org/springframework/boot/actuate/hazelcast/HazelcastHealthIndicatorServerIntegrationTests.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+/**
+ * Tests for {@link HazelcastHealthIndicator}.
+ *
+ * @author Dmytro Nosan
+ */
+class HazelcastHealthIndicatorServerIntegrationTests {
+
+	private final HazelcastInstance hazelcastServer = Hazelcast.newHazelcastInstance(new Config());
+
+	private final HazelcastHealthIndicator healthIndicator = new HazelcastHealthIndicator(this.hazelcastServer);
```
Comment (snicoll) [inline spring-boot-project/spring-boot-actuator/src/test/java/org/springframework/boot/actuate/hazelcast/HazelcastHealthIndicatorServerIntegrationTests.java]: If the plan is to write an integration test, why don't we use an `ApplicationContextRunner` with the regular auto-configuration? I wouldn't have two dedicated classes for that either.

File: spring-boot-project/spring-boot-actuator/src/test/java/org/springframework/boot/actuate/hazelcast/HazelcastHealthIndicatorTests.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+	private HazelcastInstance hazelcast;
+
+	@Mock
+	private TransactionalTaskContext context;
+
+	@Mock
+	private Client client;
+
+	@Mock
+	private Member member;
```
Comment (snicoll) [inline spring-boot-project/spring-boot-actuator/src/test/java/org/springframework/boot/actuate/hazelcast/HazelcastHealthIndicatorTests.java]: As some of these mocks are test specific, I wouldn't initialize them at class level but rather in the test that uses them.

File: spring-boot-project/spring-boot-actuator/src/test/java/org/springframework/boot/actuate/hazelcast/HazelcastHealthIndicatorServerIntegrationTests.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+	@AfterEach
+	void stop() {
+		this.hazelcastServer.shutdown();
+	}
+
+	@Test
+	void hazelcastIsUp() {
+		Health health = this.healthIndicator.health();
+		assertThat(health.getStatus()).isEqualTo(Status.UP);
```
Comment (snicoll) [inline spring-boot-project/spring-boot-actuator/src/test/java/org/springframework/boot/actuate/hazelcast/HazelcastHealthIndicatorServerIntegrationTests.java]: As this is an integration test, perhaps we could test more than up if we're paying the cost of actually bootstrapping hazelcast.

File: spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/hazelcast/HazelcastHealthIndicator.java (hunk 2 above)
Comment (nosan) [reply spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/hazelcast/HazelcastHealthIndicator.java]: This is the only way to check `Hazelcast` painless which I have found. If you have any suggestions, please let me know.

File: spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/hazelcast/HazelcastHealthIndicator.java (hunk 1 above)
Comment (nosan) [reply spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/hazelcast/HazelcastHealthIndicator.java]: Should we still split the details between `client` and `member`?

**Comment by nosan** [comment]
Thank you, @snicoll

File: spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/hazelcast/HazelcastHealthIndicator.java (hunk 1 above)
Comment (snicoll) [reply spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/hazelcast/HazelcastHealthIndicator.java]: IMO not at this time, no.

**Comment by nosan** [comment]
Thanks once again, @snicoll 

PR has been updated.

**Comment by snicoll** [comment]
Well done @nosan and thank you!

**Comment by nosan** [comment]
Thank you very much, @snicoll

**Comment by Simon3** [comment]
Hi, I see you decided to execute a null transaction on the hazelcastInstance to check its health, which is indeed good for this HealthIndicator. 

But a problem I might have with this is that it doesn't guarantee the current instance is synchronized with the cluster, i.e. all the partition migrations are done. 

I first thought I could get that behaviour by calling _isClusterSafe()_ in the HealthIndicator, but it's actually a bad idea because all instances will be considered unhealthy during any partition migration between any instances. The same problem will occur (to a lower extent) with _isMemberSafe()_. 

So I was wondering if there is any clean way to reach this behaviour, i.e. to _initially_ consider the instance unhealthy until it is in sync with the cluster.

**Comment by snicoll** [comment]
@Simon3 please do not request new feature on a closed issue. If you think the health indicator can be improved or has a problem, feel free to create a separate issue. The health indicator is about whether or not the service can be used.

**Pull request closed without merging**